### PR TITLE
Adds the functionality to notify Actors of unhandled exceptions

### PR DESCRIPTION
**Pull request by lolski**
## What is the goal of this PR?

Currently, there is a major limitation in the Actor library such that an unhandled exception in an actor instance would be swallowed by the event loop.

This PR adds the functionality allowing actors to be notified of the exception that it failed to handle. This would allow actors to decide on whether to recover (if said actor considers the exception non-fatal) or shut down (if it considers the exception fatal).

```
class MyActor extends Actor<MyActor> {
  @Override
  public void exception(Exception e) {
    // handle the uncaught exception
  }
}
```

This PR depends on: https://github.com/graknlabs/common/pull/47.

### CODE REVIEW

File: concurrent/actor/eventloop/EventLoop.java [hunk 1]
```diff
@@ -45,52 +49,60 @@ public EventLoop(ThreadFactory factory) {
...
+                        } catch (Exception e) {
+                            event.second().accept(e);
+                        }
                     }
                 }
             }
-        } catch (Exception ex) {
-            LOG.error("EventLoop error", ex);
+        } catch (Exception e) {
+            errorHandler.accept(e);
```
Comment (flyingsilverfin) [inline concurrent/actor/eventloop/EventLoop.java:91]: this looks odd - all error handling is now happening inside the `while` loop, so we shouldn't have to have any sophisticated error handling outside the loop

**Review by flyingsilverfin — APPROVED** [review]
looks good pending 1 comment

File: concurrent/actor/eventloop/EventLoop.java (hunk 1 above)
Comment (lriuui0x0) [reply concurrent/actor/eventloop/EventLoop.java:91]: agreed - does code outside the `while` loop can create error?

File: concurrent/actor/eventloop/EventLoop.java [hunk 2]
```diff
@@ -32,9 +35,10 @@
 public class EventLoop {
     private static final Logger LOG = LoggerFactory.getLogger(EventLoop.class);
 
-    private final TransferQueue<Runnable> eventQueue = new LinkedTransferQueue<>();
-    private final LogicalTimerQueue<Runnable> timerQueue = new LogicalTimerQueue<>();
+    private final TransferQueue<Pair<Runnable, Consumer<Exception>>> events = new LinkedTransferQueue<>();
+    private final LogicalTimerQueue<Runnable> timers = new LogicalTimerQueue<>();
```
Comment (lriuui0x0) [inline concurrent/actor/eventloop/EventLoop.java:39]: `timers` need to be able to handle errors as well.

Comment (lolski) [reply concurrent/actor/eventloop/EventLoop.java:39]: I'm thinking of what should we do when a timer task throws an exception. Should we just trigger the `exception` method of the Actor the same way we do for normal message?

File: concurrent/actor/eventloop/EventLoop.java (hunk 1 above)
Comment (lolski) [reply concurrent/actor/eventloop/EventLoop.java]: Fixed

File: concurrent/actor/eventloop/EventLoop.java (hunk 2 above)
Comment (lolski) [reply concurrent/actor/eventloop/EventLoop.java:39]: I've agreed with @lriuui0x0 to take care of the issue on a separate PR.

**Pull request merged**